Support look-back for nest BMI module input values and bootstrapping of defaults.

robertbartel opened the pull request:
Add support for allowing a nested formulation in a BMI multi formulation to use as a provider a later nested formulation.  Previously, only earlier nested formulations could be used like this, where earlier/later is with respect to their listed order in the realization config.  Since a later nested formulation's outputs won't have been (re)calculated yet, this effectively is a look-back to the previous time step.

Also, adding support for configuring a default value for nested formulation outputs, including having it be used as the value during the first time step in the aforementioned look-back scenario.

hellkite500 commented on include/forcing/OptionalWrappedProvider.hpp:
```diff
@@ -0,0 +1,361 @@
+            // Also, in this implementation, don't count usages until there is a backing provider that can provide this
+            auto waits_it = defaultUsageWaits.find(output_name);
+            if (waits_it != defaultUsageWaits.end() && isSuppliedByWrappedProvider(output_name)) {
+                // A value of 0 should be inferred and cleared from the collection, so ...
```
So what exactly happens with a value of 0???

robertbartel replied:
@hellkite500, the intent was for there never be a value less than `1` in the `defaultUsageWaits` map.  If we assume this condition when an object is created, then it should hold throughout, as any values equal to 1 before the decrementing step are instead removed from the map entirely.

However, your comment prompted me to double-check and discover that it is possible for the `> 0` condition to be violated.  There is this constructor: 

```c++
OptionalWrappedProvider(vector<string> providedOuts, map<string, double> defaultVals, map<string, int> defaultWaits)
```

This validates the keys in `defaultWaits` (needed for other reasons) but not that the values are all greater than 0.

I could fix this, but first I'd like some opinions on which is the right approach.  Instead of what I was originally trying, does it make more sense to do as you had expected and check in the `recordUsingDefault` function for the `<= 0` case?  It's slightly less efficient but more intuitive.

hellkite500 commented on include/realizations/catchment/Bmi_Multi_Formulation.hpp:
```diff
@@ -506,6 +508,50 @@ namespace realization {
+         *
+         * During nested formulation creation, when a nested formulation requires as input some output expected from
+         * soon-to-be-created (i.e., later in execution order) formulation (e.g., in a look-back scenario to an earlier
+         * time step), then a deferred provider gets registered with the nested module and has a referenced added to
```
"and has a reference added to"

hellkite500 commented:
A couple additional thoughts.  Should the `default`/`init` value not found in configuration attempt to use the `bmi` `get_value` to query the formulation for a value?

Would we ever want more than a single time step "look back"?

mattw-nws commented:
This didn't make AGU/v0.1.0 but I have a feeling it will be wanted for "MVP" to support [SoilFreezeThaw](/NOAA-OWP/SoilFreezeThaw) along with #343 .

robertbartel commented:
>Should the default/init value not found in configuration attempt to use the bmi get_value to query the formulation for a value?

@hellkite500, I'm not sure I understand exactly what you are asking, though it does seem as though perhaps this may not sufficiently handle the case when a default is needed but not provided.  I'll have to double-check that, but I'd like to make sure I understand what you were suggesting before I do.

>Would we ever want more than a single time step "look back"?

Possibly, but that has the potential to be a considerably more significant undertaking.  E.g., how far back do we save?  How do we control what is saved beyond one time step?  Where/how is data being saved?  Are we doing anything to ensure we don't exceed storage capacity?  

Unless there is something specific that is fairly simple/similar and immediately needed, we should handle expanding this functionality separately.  If there are any particular use cases that we have need for, we'll also need those details.

robertbartel commented:
>the intent was for there never be a value less than 1 in the defaultUsageWaits map. If we assume this condition when an object is created, then it should hold throughout, as any values equal to 1 before the decrementing step are instead removed from the map entirely.

@hellkite500, I realized after our discussion last week, as I was working on fixing the "problem" with this, that I been wrong it how I remembered this was supposed to work.   But I think we need to change things anyway.

In short, things were made to support perpetually providing a default value for some output.  The 0-value case you originally asked about was to allow this, as was the possibility of the constructor receiving wait values less than 1.

Or, stated otherwise, the intent was actually for there never be a value less than 1 in the `defaultUsageWaits` map, _except_ when we would not want a wait value to be altered or removed.  And the latter would be for values indicating that we wanted to perpetually use a default value.

After our talk, I don't think this should be handled by the framework, so I've modified things to take that out.  It would have changed the nature of the type from being a wrapper/proxy to being a data source.

Also,`setup_nested_deferred_provider` in _Bmi_Multi_Formulation_ is written so that these deferred providers will not have an initialized default if none is provided via the realization config.  And I've added some unit tests to make sure the wrapper falls back to the back to the value of the source provider if no default is given.

robertbartel commented:
Added some test case that were not covered after conversation with @mattw-nws.  Also refactored the test class a bit.  Plus, rebased against latest `master`.

mattw-nws commented:
This (unit conversion stderr text removed:

```
/Users/runner/work/ngen/ngen/test/realizations/catchments/Bmi_Multi_Formulation_Test.cpp:653: Failure
Expected equality of these values:
  expected
    Which is: 4.86646e-08
  response
    Which is: 2.78098e-08
[  FAILED  ] Bmi_Multi_Formulation_Test.GetResponse_3_b (220 ms)
```

Looks like you duplicated test `GetResponse_0_b` before f5b33bc, you'll want to propagate the changes to the values in that commit.

mattw-nws commented:
Merging, but noting possible gotcha: I don't see a test that makes sure reading `default_output_values` through from the realization config works. Testing the use of default vals is done via providing defaults direct via C++ API, though. Just keep an eye on whether it works when doing this via the realization config (wouldn't worry about it except this bit me in #343!).
